axlearn/common/checkpointer_orbax_emergency.py

```diff
@@ -0,0 +1,814 @@
+        # restore from either to succeed.
+        def _composite_save_policy(*, step: int, evaler_summaries: dict[str, Any]):
+            return (
+                save_policy(step=step, evaler_summaries=evaler_summaries)
+                or local_save_policy(step=step, evaler_summaries=evaler_summaries)
+                or self._reached_preemption
+            )
+
+        ckpt_cfg.save_policy = config_for_function(lambda: _composite_save_policy)
+        self._non_tensor_manager: Checkpointer = ckpt_cfg.instantiate(parent=self)
+        self._tensor_manager: Optional[oecp.CheckpointManager] = None
+        # See comments of _eval_summaries in `OrbaxCheckpointer`.
+        self._eval_summaries = None
+        self._reached_preemption = False
+
+    # pylint: disable-next=redefined-builtin
+    def ckpt_dir(self, step: int, dir: Optional[str] = None) -> str:
+        """Obtains the checkpoint dir for the given step."""
+        if dir is None:
+            dir = self._non_tensor_manager.directory
+        return str(ocp.step.build_step_path(dir, self._name_format, step))
+
+    def _get_abstract_state(
+        self, state_with_tensors: Nested[Tensor]
+    ) -> Nested[jax.ShapeDtypeStruct]:
+        """Generate the abstract states required by the Orbax emergency checkpointer."""
+        return jax.tree.map(
+            lambda x: jax.ShapeDtypeStruct(x.shape, x.dtype, sharding=x.sharding),
+            state_with_tensors,
+        )
+
+    def _get_tensor_manager(self, state_with_tensors: Nested[Tensor]) -> oecp.CheckpointManager:
+        """Creates the emergency checkpoint manager if not exists.
+
+        We defer the creation of this checkpoint manager because it requires the state dict,
+        which is not present during __init__.
+        """
+        cfg: OrbaxEmergencyCheckpointer.Config = self.config
+        if self._tensor_manager is not None:
+            return self._tensor_manager
+
+        save_policy = cfg.save_policy.instantiate()
+        local_save_policy = cfg.local_save_policy.instantiate()
+
+        def _orbax_save_fn(
+            step: int, last_saved_step: Optional[int], wrapped_save_policy: CheckpointPolicy
+        ) -> bool:
+            del last_saved_step
+            return wrapped_save_policy(step=step, evaler_summaries=self._eval_summaries)
+
+        # For meaning of these options, refer to
+        # https://github.com/google/orbax/blob/95be2c021bc8cbf4badd83a053ff57b7a9f9b314/checkpoint/orbax/checkpoint/experimental/emergency/checkpoint_manager.py#L277
+        self._tensor_manager = oecp.CheckpointManager(
+            self._local_dir,
+            persistent_directory=os.path.join(cfg.dir, self._TENSORS_PREFIX),
+            global_mesh=thread_resources.env.physical_mesh,
+            abstract_state=self._get_abstract_state(state_with_tensors),
+            options=oecp.CheckpointManagerOptions(
+                local=oecp.LocalCheckpointOptions(
```

Comment:
Shall we expose `oecp.LocalCheckpointOptions` to users as `Config.local_checkpoint_options`? User can set it to None to disable local checkpoints.

We can provide a helper function for users to construct `should_save_fn` from their policy.

Reply:
It's already exposed, via `local_keep_last_n` and `local_save_policy`